packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -0,0 +1,363 @@
+   * The `visibleCount` is not changed if the correct `visibleCount` has already been found.
+   *
+   * Logic:
+   * - Have a guess range for `visibleCount`. e.g. `[0, 32]` (32 is an arbitrary choice).
+   * - Keep doubling the max guess until the container overflows.
+   *   i.e. the max guess should always be `≥` the correct `visibleCount`.
+   *   - With each such doubling, the new min guess is the current max guess (since underflow = we guessed low).
+   * - Set `visibleCount` to the `maxGuess`.
+   * - Repeat the following by calling `guessVisibleCount()` (keep re-rendering but not painting):
+   *   - Each time the container overflows, new max guess is the average of the two guesses.
+   *   - Each time the container does not overflow, new min guess is the average of the two guesses.
+   * - Stop when the average of the two guesses is the min guess itself. i.e. no more averaging possible.
+   * - The min guess is then the correct `visibleCount`.
+   */
+  const guessVisibleCount = React.useCallback(() => {
+    // If disabled or already stabilized
+    if (disabled || isStabilized || isGuessing.current) {
+      return;
+    }
+
+    // We need to wait for the ref to be attached so that we can measure available and required sizes.
+    if (containerRef.current == null) {
+      return;
+    }
+
+    try {
+      isGuessing.current = true;
+
+      const dimension = orientation === 'horizontal' ? 'Width' : 'Height';
+      const availableSize = containerRef.current[`offset${dimension}`];
+      const requiredSize = containerRef.current[`scroll${dimension}`];
+
+      const isOverflowing = availableSize < requiredSize;
+
+      // We have already found the correct visibleCount
+      if (
+        (visibleCount === itemsLength && !isOverflowing) ||
+        // if the new average of visibleCountGuessRange will never change the visibleCount anymore (infinite loop)
+        (visibleCountGuessRange[1] - visibleCountGuessRange[0] === 1 &&
+          visibleCount === visibleCountGuessRange[0])
+      ) {
+        setVisibleCountGuessRange(null);
+        return;
+      }
+
+      // Before the main logic, the max guess MUST be >= the correct visibleCount for the algorithm to work.
+      // If not:
+      // - double the max guess and visibleCount: since we need to overflow.
+      // - set min guess to current visibleCount: since underflow means correct visibleCount >= current visibleCount.
+      if (visibleCountGuessRange[1] === visibleCount && !isOverflowing) {
+        const doubleOfMaxGuess = visibleCountGuessRange[1] * 2;
+
+        setVisibleCountGuessRange([visibleCount, doubleOfMaxGuess]);
+        setVisibleCount(doubleOfMaxGuess);
+
+        return;
+      }
+
+      let newVisibleCountGuessRange = visibleCountGuessRange;
+
+      if (isOverflowing) {
+        // overflowing = we guessed too high. So, new max guess = half the current guess
+        newVisibleCountGuessRange = [visibleCountGuessRange[0], visibleCount];
```

Comment:
I don't think that I understand how this part of the algorithm works. This is apparently halving the `visibleCountGuesssRange`, but I don't know how `visibleCount` is half of the max for the guess range. I must be missing something, since I am able to test the overflow container out with the deploy preview and see that it is working.